### PR TITLE
feat(sync): #MC-171 sync external events from url

### PR DESCRIPTION
**Pull request by alicedraillard**
## Describe your changes
Allow to synchronize an external calendar (previously imported for the first time). **All events since the last update** are deleted and then recreated. 

https://confluence.support-ent.fr/display/MAgenda/Calendar+API

## Checklist tests
Call API using the id of an external event

## Issue ticket number and link
[MC-171](https://jira.support-ent.fr/browse/MC-171)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/net/atos/entng/calendar/controllers/CalendarController.java [hunk 1]
```diff
@@ -209,9 +209,26 @@ Future<JsonObject> createFuture(UserInfos user, JsonObject body) {
...
+            if (user != null) {
+                String host = getHost(request);
+                String i18nLang = I18n.acceptLanguage(request);
+                calendarHelper.externalCalendarSync(calendarId,
+                                user, host, i18nLang, ExternalICalEventBusActions.PUT.method())
+                        .onSuccess(result -> {
+                            Renders.ok(request);
+                        })
+                        .onFailure(error -> {
+                            String message = String.format("[Calendar@%s::importExternalCalendar] An error has occured" +
```
Comment (flmartineau) [inline src/main/java/net/atos/entng/calendar/controllers/CalendarController.java:223]: ::syncExternalCalendar

File: src/main/java/net/atos/entng/calendar/helpers/CalendarHelper.java [hunk 2]
```diff
@@ -5,68 +5,119 @@
...
+            return Future.failedFuture(message);
+        }
+
+        JsonObject calendar = calendarList.getJsonObject(0);
+        switch(action) {
+            case Field.POST:
+                updateExternalCalendar(params, calendar, true)
+                        .onSuccess(promise::complete)
+                        .onFailure(error -> {
+                            String message = String.format("[Calendar@%s::prepareCalendarAndEventsForUpdate]:  an error has occurred while " +
```
Comment (flmartineau) [inline src/main/java/net/atos/entng/calendar/helpers/CalendarHelper.java:79]: Tu peux inclure le error dans ton log peut-être ?

File: src/main/java/net/atos/entng/calendar/services/impl/EventServiceMongoImpl.java [hunk 3]
```diff
@@ -290,6 +312,29 @@ public Future<JsonObject> delete(String calendarId, String eventId, UserInfos us
...
+    public Future<Void> deleteDatesAfterComparisonDate(String calendarId, String comparisonDate) {
+        Promise<Void> promise = Promise.promise();
+
+        // Query
+        QueryBuilder query = QueryBuilder.start("calendar").is(calendarId);
+        query.put(Field.STARTMOMENT).greaterThan(comparisonDate);
+
+        mongo.delete(this.collection, MongoQueryBuilder.build(query), result -> {
+            if (result.body().isEmpty()) {
+                String message = String.format("[Calendar@%s::deleteDatesAfterDate]:  " +
```
Comment (flmartineau) [inline src/main/java/net/atos/entng/calendar/services/impl/EventServiceMongoImpl.java:325]: ::deleteDatesAfterComparisonDate

File: src/main/java/net/atos/entng/calendar/helpers/CalendarHelper.java [hunk 4]
```diff
@@ -5,68 +5,119 @@
...
+                            return eventServiceMongo.deleteDatesAfterComparisonDate(calendar.getString(Field._ID), DateUtils.dateToString(lastUpdateDate));
+                        })
+                        .compose(result -> updateExternalCalendar(params, calendar, true))
+                        .onSuccess(promise::complete)
+                        .onFailure(error -> {
+                            String message = String.format("[Calendar@%s::prepareCalendarAndEventsForUpdate]:  an error has occurred while " +
+                                            "preparing calendar and event for sync: %s",
+                                    this.getClass().getSimpleName(), error.getMessage());
+                            log.error(message);
+                            Future.failedFuture(message);
```
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/helpers/CalendarHelper.java]: Linker à la promise

File: src/main/java/net/atos/entng/calendar/ical/ExternalImportICal.java [hunk 5]
```diff
@@ -69,7 +72,29 @@ private void createEventsFromICal(Message<JsonObject> message, UserInfos user) {
         this.requestInfo = message.body().getJsonObject(Field.REQUEST, new JsonObject());
 
         fetchICalFromUrl(message.body().getJsonObject(Field.CALENDAR, new JsonObject()).getString(Field.ICSLINK, null))
-                .compose(ical -> eventService.importIcal(calendar.getString(Field._ID), ical, user, requestInfo, Field.CALENDAREVENT))
+                .compose(ical -> eventService.importIcal(calendar.getString(Field._ID), ical, user, requestInfo,
+                        Field.CALENDAREVENT))
+                .onSuccess(message::reply)
+                .onFailure(err -> {
+                    String errMessage = String.format("[Calendar@%s::fetchICalFromUrl]:  " +
```
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/ical/ExternalImportICal.java:79]: Nom de méthode

File: src/main/java/net/atos/entng/calendar/ical/ExternalImportICal.java [hunk 6]
```diff
@@ -69,7 +72,29 @@ private void createEventsFromICal(Message<JsonObject> message, UserInfos user) {
...
+        this.requestInfo = message.body().getJsonObject(Field.REQUEST, new JsonObject());
+        String calendarLastUpdate = new Date(this.calendar.getJsonObject(Field.UPDATED, new JsonObject()).getLong("$date"))
+                .toInstant().toString();
+
+        fetchICalFromUrl(message.body().getJsonObject(Field.CALENDAR, new JsonObject()).getString(Field.ICSLINK, null))
+                .compose(ical -> eventService.importIcal(calendar.getString(Field._ID), ical, user, requestInfo,
+                        Field.CALENDAREVENT, ExternalICalEventBusActions.SYNC.method(), calendarLastUpdate))
                 .onSuccess(message::reply)
                 .onFailure(err -> {
                     String errMessage = String.format("[Calendar@%s::fetchICalFromUrl]:  " +
```
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/ical/ExternalImportICal.java:100]: Same

File: src/main/java/net/atos/entng/calendar/ical/ICalHandler.java [hunk 7]
```diff
@@ -65,6 +65,8 @@ public class ICalHandler extends AbstractVerticle implements Handler<Message<Jso
     public static final String ACTION_PUT = "put";
     public static final String ACTION_GET = "get";
 
+    public static final String ACTION_SYNC = "sync";
```
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/ical/ICalHandler.java:68]: Enum

File: src/main/java/net/atos/entng/calendar/ical/ICalHandler.java [hunk 8]
```diff
@@ -153,7 +163,8 @@ private void icsContentToJsonEvents(Message<JsonObject> message) {
                         setEventDates(event, jsonEvent);
                         setEventProperties(event, jsonEvent, requestInfo);
                         validateEvent(event);
-                        events.add(jsonEvent);
+                        if((lastUpdate == null) || DateUtils.parseDate(jsonEvent.getString(Field.STARTMOMENT), DateUtils.DATE_FORMAT_UTC)
+                                .after(DateUtils.parseDate(lastUpdate, DateUtils.DATE_FORMAT_UTC))) events.add(jsonEvent);
                     } catch (UnhandledEventException uee) {
                         jsonEvent.put("errorCause", uee.getMessage());
```
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/ical/ICalHandler.java:169]: Magic string

File: src/main/java/net/atos/entng/calendar/services/impl/EventServiceMongoImpl.java (hunk 3 above)
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/services/impl/EventServiceMongoImpl.java:325]: method name

**Review by flmartineau — APPROVED** [review]

File: src/main/java/net/atos/entng/calendar/services/impl/EventServiceMongoImpl.java [hunk 9]
```diff
@@ -290,6 +312,29 @@ public Future<JsonObject> delete(String calendarId, String eventId, UserInfos us
         return promise.future();
     }
 
+    @Override
+    public Future<Void> deleteDatesAfterComparisonDate(String calendarId, String comparisonDate) {
+        Promise<Void> promise = Promise.promise();
+
+        // Query
+        QueryBuilder query = QueryBuilder.start("calendar").is(calendarId);
```
Comment (qperie) [inline src/main/java/net/atos/entng/calendar/services/impl/EventServiceMongoImpl.java:320]: Magic string ?

File: src/main/java/net/atos/entng/calendar/helpers/CalendarHelper.java [hunk 10]
```diff
@@ -5,68 +5,120 @@
...
+                                            "preparing calendar for first sync: %s",
+                                    this.getClass().getSimpleName(), error.getMessage());
+                            log.error(message);
+                            promise.fail(message);
+                        });
+                break;
+            case Field.PUT:
+                eventServiceMongo.retrieveByCalendarId(calendar.getString(Field._ID))
+                        .compose(eventList -> {
+                            Timestamp lastUpdateTimestamp = new Timestamp(calendar.getJsonObject(Field.UPDATED).getLong("$date"));
```
Comment (BastienBenard) [inline src/main/java/net/atos/entng/calendar/helpers/CalendarHelper.java:90]: Je ne sais pas si cela existe sur calendar mais tu peux crée un MongoField pour "$date"

**Review by qperie — APPROVED** [review]

File: src/main/java/net/atos/entng/calendar/core/constants/Field.java [hunk 11]
```diff
@@ -136,8 +137,10 @@ private Field() {
     public static final String URL = "url";
     public static final String ISEXTERNAL = "isExternal";
     public static final String ICSLINK = "icsLink";
+    public static final String SYNC = "sync";
 
     //Http Methods
     public static final String POST = "post";
     public static final String PUT = "put";
+
 }
```
Comment (aymericdevriesere) [inline src/main/java/net/atos/entng/calendar/core/constants/Field.java:146]: No new line

**Review by aymericdevriesere — APPROVED** [review]

**Pull request merged**